allennlp/nn/beam_search.py

```diff
@@ -763,7 +865,12 @@ def _search(
         # shape: (batch_size, beam_size, max_steps)
         all_predictions = torch.cat(list(reversed(reconstructed_predictions)), 2)
 
-        return all_predictions, last_log_probabilities
+        # Calculate the final sequence scores
+        final_scores = self.final_sequence_scorer.score(
+            all_predictions, last_log_probabilities, self._end_index
+        )
```

Comment:
We should re-sort at this point to make sure the highest scoring sequences are ordered first. A lot of downstream code depends on this.

```suggestion
        final_scores = self.final_sequence_scorer.score(
            all_predictions, last_log_probabilities, self._end_index
        )
```

Reply:
Added this change and a unit test that uses a length penalty to reverse the original order of the sequences to ensure the returned sequences are in the new order